### PR TITLE
Insert items and relationship error/timeout handling

### PR DESCRIPTION
**Pull request by RickyV33**
Added atomic functionality to graph constructor. If it errors with a timeout when its requesting the name, items, or relationship, then none of the graph fields are written to.

### CODE REVIEW

File: lib/graph.js [hunk 1]
```diff
@@ -16,20 +17,63 @@ class Graph {
...
+          return;
+        }
+        ++counter;
+        //  ----------------------
+        // | Getting project name |
+        //  ----------------------
+        projectquire.getProjectName(projectId, url).then(name => {
+          this.name = name;
+        }, error => {
+          if (error.code === 'ETIMEDOUT' || error.connect === true) {
```
Comment (ruv-prog-so) [inline lib/graph.js]: The story says that the if there is an error, then we should not populate the graph object. This code says that if there is an error, then get the data.

File: lib/graph.js [hunk 2]
```diff
@@ -16,20 +17,63 @@ class Graph {
...
+        projectquire.getProjectItems(projectId, url).then(items => {
+          this.nodes = items;
+        }, error => {
+          if (error.code === 'ETIMEDOUT' || error.connect === true) {
+            getData();
+          } else {
+            console.error('Request Error on retrieving project items: ' + error.statusCode + ' ' + error.status);
+          }
+        });
+        //  -----------------------
```
Comment (HansenML) [inline lib/graph.js]: Looks like the conditional is backwards here. If you have an error, you should not be getting the items, you should be outputting the message. I think the if/else needs to be inverted.

File: lib/graph.js [hunk 3]
```diff
@@ -16,20 +17,63 @@ class Graph {
...
+            console.error('Request Error on retrieving project name: ' + error.statusCode + ' ' + error.status);
+          }
+        });
+        //  ---------------
+        // | Getting items |
+        //  ---------------
+        projectquire.getProjectItems(projectId, url).then(items => {
+          this.nodes = items;
+        }, error => {
+          if (error.code === 'ETIMEDOUT' || error.connect === true) {
```
Comment (HansenML) [inline lib/graph.js]: If error.code has your timeout value and error.connect is true, then you should not be gathering the items. Looks like you need to invert your if/else here.

File: lib/graph.js [hunk 4]
```diff
@@ -16,20 +17,63 @@ class Graph {
...
+        });
+        //  -----------------------
+        // | Getting relationships |
+        //  -----------------------
+        projectquire.getProjectRelationships(projectId, url).then(
+          relationships => {
+            this.edges = relationships;
+          },
+          error => {
+            if (error.code === 'ETIMEDOUT' || error.connect === true) {
```
Comment (HansenML) [inline lib/graph.js]: Same issue as the other if/else statements. Looks like you need to invert the conditionals.

File: lib/pagination.js [hunk 5]
```diff
@@ -24,9 +24,13 @@ function pagination (baseUrl, startAt, maxResults) {
           url = baseUrl + '?' + startAtParam + '&' + maxResultsParam;
         }
         request({url: url, json: true}, function (error, response, body) {
-          if (response.body.meta.status !== 'OK' || error) {
+          if (response.statusCode !== 200 || error) {
```
Comment (ruv-prog-so) [inline lib/pagination.js]: What if the status code is another 200 level code returned? For example, what if you get at 201 status, which means accepted? It should not return an error.
https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

File: test/lib/graph.js [hunk 6]
```diff
@@ -64,21 +64,97 @@ describe('Graph class module', () => {
         expect(newGraph.edges[0].fromItem).to.equal(1);
         expect(newGraph.edges[0].toItem).to.equal(2);
         expect(newGraph.edges[0].type).to.equal(99);
-      }, 2000);
+      }, 5000);
+    });
+
+    it('should return a graph that contains nothing when either project ID or url are invalid.', () => {
```
Comment (HansenML) [inline test/lib/graph.js]: The first and last test cases here are both for the case that all three are rejected. It looks like you forgot to test for the valid case. So I would change one of the duplicate "all rejected" cases into an "all valid" case. Might be helpful to use a truth table to iterate through for large test blocks like this. @ruv-prog-so and I only caught this error by making a truth table to check your cases against.

File: lib/graph.js (hunk 1 above)
Comment (RickyV33) [inline lib/graph.js]: So this is the functionality that makes it atomic. If we encounter an error, then we are supposed to retry again 5 times. I think we just need to rename the getData function to make that a little bit more clear. If that condition is true, then there was a timeout, and it's going to run the nested function, getData(), again and increment a counter. Once getData has been called 5 times, it logs an error message and doesn't retrieve any data. This call happens at either the retrieval of project name, items, or relationships.

**Pull request merged**